File: src/GroqClient.php
<?php

declare(strict_types=1);

use OpenAI\Exceptions\ErrorException;

final class GroqClient
{
    private const MAX_ITERATIONS = 10;
    private const MAX_TOOL_JSON_BYTES = 14000;
    private const MAX_429_RETRIES = 5;

    private \OpenAI\Client $client;

    public function __construct(
        private string $apiKey,
        private string $model = 'llama-3.1-8b-instant'
    ) {
        $this->client = \OpenAI::factory()
            ->withApiKey($apiKey)
            ->withBaseUri('api.groq.com/openai/v1')
            ->make();
    }

    /**
     * @param list<array<string, mixed>> $messages
     * @param list<array<string, mixed>> $tools
     * @param callable(string, array): string $runTool nombre función + args decodificados -> JSON string resultado
     * @return array{reply: string, messages: list<array<string, mixed>>}
     */
    public function chatWithTools(array $messages, array $tools, callable $runTool): array
    {
        $working = $messages;
        $iteration = 0;

        while ($iteration < self::MAX_ITERATIONS) {
            $iteration++;
            $response = $this->requestCompletion($working, $iteration === 1 ? $tools : []);

            $choice = $response->choices[0] ?? null;
            if ($choice === null) {
                return ['reply' => 'Respuesta vacía del modelo.', 'messages' => $working];
            }

            $message = $choice->message;
            $toolCalls = $message->toolCalls ?? [];

            $assistantPayload = [
                'role'    => 'assistant',
                'content' => $message->content,
            ];
            if (!empty($toolCalls)) {
                $assistantPayload['tool_calls'] = array_map(
                    fn($tc) => [
                        'id'       => $tc->id,
                        'type'     => $tc->type,
                        'function' => [
                            'name'      => $tc->function->name,
                            'arguments' => $tc->function->arguments,
                        ],
                    ],
                    $toolCalls
                );
            }
            $working[] = $assistantPayload;

            if (empty($toolCalls)) {
                return ['reply' => (string) ($message->content ?? ''), 'messages' => $working];
            }

            foreach ($toolCalls as $tc) {
                $args = json_decode($tc->function->arguments, true);
                if (!is_array($args)) {
                    $args = [];
                }
                $toolJson = $this->clampToolJson($runTool($tc->function->name, $args));
                $working[] = [
                    'role'         => 'tool',
                    'tool_call_id' => $tc->id,
                    'content'      => $toolJson,
                ];
            }
        }

        return ['reply' => 'Se alcanzó el límite de iteraciones con herramientas.', 'messages' => $working];
    }

    private function requestCompletion(array $messages, array $tools): \OpenAI\Responses\Chat\CreateResponse
    {
        $params = [
            'model'       => $this->model,
            'messages'    => $messages,
            'temperature' => 0.2,
        ];
        if ($tools !== []) {
            $params['tools']       = $tools;
            $params['tool_choice'] = 'auto';
        }

        for ($attempt = 1; $attempt <= self::MAX_429_RETRIES; $attempt++) {
            try {
                return $this->client->chat()->create($params);
            } catch (ErrorException $e) {
                $msg = $e->getMessage();

                if ($this->isGroqDailyTokenLimit($msg)) {
                    throw new \RuntimeException($this->groqFriendlyDailyLimitMessage($msg));
                }

                if ($this->isRateLimit($msg) && $attempt < self::MAX_429_RETRIES) {
                    $sleepSec = $this->parseGroqRetrySeconds($msg);
                    if ($sleepSec < 1.0) {
                        $sleepSec = min(8.0, 2.0 * $attempt);
                    }
                    usleep((int) round($sleepSec * 1_000_000));
                    continue;
                }

                throw new \RuntimeException($this->groqFriendlyErrorMessage($msg));
            }
        }

        throw new \RuntimeException('Groq: se agotaron los reintentos sin obtener una respuesta válida.');
    }

    private function isGroqDailyTokenLimit(string $message): bool
    {
        return stripos($message, 'tokens per day') !== false
            || stripos($message, 'TPD') !== false;
    }

    private function isRateLimit(string $message): bool
    {
        return stripos($message, 'rate_limit') !== false
            || stripos($message, 'rate limit') !== false
            || stripos($message, 'too many requests') !== false;
    }

    private function groqFriendlyDailyLimitMessage(string $raw): string
    {
        $hint = '';
        if (preg_match('/try again in\s+([^\n.]+)/i', $raw, $m)) {
            $hint = ' Indicación de Groq: esperar ~' . trim($m[1]) . '.';
        }
        return 'Se alcanzó el límite diario de tokens (TPD) de Groq para este modelo.' . $hint
            . ' Reduce uso (menos preguntas repetidas, menos historial en el chat), prueba otro modelo en GROQ_MODEL (.env), o revisa tu cuota en console.groq.com.'
            . ' [Detalle] ' . $raw;
    }

    private function groqFriendlyErrorMessage(string $raw): string
    {
        if ($this->isRateLimit($raw)) {
            return 'Groq devolvió límite de velocidad (429). Espera unos minutos o cambia de modelo en GROQ_MODEL. [Detalle] ' . $raw;
        }
        return 'Groq error: ' . $raw;
    }

    private function parseGroqRetrySeconds(string $message): float
    {
        if (preg_match('/try again in ([\d.]+)\s*s/i', $message, $m)) {
            return min(30.0, max(0.5, (float) $m[1]));
        }
        return 0.0;
    }

    private function clampToolJson(string $json): string
    {
        if (strlen($json) <= self::MAX_TOOL_JSON_BYTES) {
            return $json;
        }
        $data = json_decode($json, true);
        $sliceKeys = ['filas', 'filas_ranking', 'filas_pareto'];
        foreach ($sliceKeys as $key) {
            if (!is_array($data) || !isset($data[$key]) || !is_array($data[$key])) {
                continue;
            }
            $all = $data[$key];
            foreach ([12, 8, 5, 3] as $n) {
                $data[$key] = array_slice($all, 0, $n);
                $data['_nota'] = "Respuesta truncada: primeras {$n} filas en {$key} (límite API).";
                $out = json_encode($data, JSON_UNESCAPED_UNICODE | JSON_INVALID_UTF8_SUBSTITUTE);
                if ($out !== false && strlen($out) <= self::MAX_TOOL_JSON_BYTES) {
                    return $out;
                }
            }
            $data[$key] = $all;
        }
        return json_encode([
            'error'   => 'respuesta_herramienta_muy_grande',
            'mensaje' => 'La consulta devolvió demasiados datos. Acote fechas o filtros.',
        ], JSON_UNESCAPED_UNICODE);
    }
}

File: api/python/app/groq_client.py
from __future__ import annotations

import json
import re
import time
from typing import Any, Callable

from openai import OpenAI

MAX_ITERATIONS = 10
MAX_TOOL_JSON_BYTES = 14000
MAX_429_RETRIES = 3
_MAX_WAIT_SEC = 25.0


class GroqClient:
    def __init__(self, api_key: str, model: str = "llama-3.1-8b-instant") -> None:
        self._client = OpenAI(api_key=api_key, base_url="https://api.groq.com/openai/v1")
        self._model = model

    def chat_with_tools(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]],
        run_tool: Callable[[str, dict[str, Any]], str],
    ) -> dict[str, Any]:
        working = list(messages)
        iteration = 0
        while iteration < MAX_ITERATIONS:
            iteration += 1
            use_tools = tools if iteration == 1 else []
            response = self._request_completion(working, use_tools)
            choice = response.choices[0]
            msg = choice.message
            tool_calls = getattr(msg, "tool_calls", None) or []

            assistant_payload: dict[str, Any] = {
                "role": "assistant",
                "content": msg.content,
            }
            if tool_calls:
                assistant_payload["tool_calls"] = [
                    {
                        "id": tc.id,
                        "type": tc.type,
                        "function": {"name": tc.function.name, "arguments": tc.function.arguments},
                    }
                    for tc in tool_calls
                ]
            working.append(assistant_payload)

            if not tool_calls:
                return {"reply": msg.content or "", "messages": working}

            for tc in tool_calls:
                try:
                    args = json.loads(tc.function.arguments or "{}")
                except json.JSONDecodeError:
                    args = {}
                if not isinstance(args, dict):
                    args = {}
                tool_json = self._clamp_tool_json(run_tool(tc.function.name, args))
                working.append({"role": "tool", "tool_call_id": tc.id, "content": tool_json})

        return {
            "reply": "Se alcanzó el límite de iteraciones con herramientas.",
            "messages": working,
        }

    def _request_completion(
        self, messages: list[dict[str, Any]], tools: list[dict[str, Any]]
    ) -> Any:
        params: dict[str, Any] = {
            "model": self._model,
            "messages": messages,
            "temperature": 0.2,
        }
        if tools:
            params["tools"] = tools
            params["tool_choice"] = "auto"

        last_err: Exception | None = None
        for attempt in range(1, MAX_429_RETRIES + 1):
            try:
                return self._client.chat.completions.create(**params)
            except Exception as e:
                last_err = e
                msg = str(e)
                if self._is_groq_daily_token_limit(msg):
                    raise RuntimeError(self._friendly_daily(msg)) from e
                if self._is_rate_limit(msg) and attempt < MAX_429_RETRIES:
                    sleep_sec = self._get_wait_seconds(e, msg)
                    if sleep_sec > _MAX_WAIT_SEC:
                        raise RuntimeError(self._friendly_wait_long(sleep_sec)) from e
                    time.sleep(sleep_sec)
                    continue
                raise RuntimeError(self._friendly_error(msg)) from e
        raise RuntimeError(str(last_err) if last_err else "Groq: sin respuesta")

    @staticmethod
    def _is_groq_daily_token_limit(message: str) -> bool:
        m = message.lower()
        return "tokens per day" in m or "tpd" in m

    @staticmethod
    def _is_rate_limit(message: str) -> bool:
        m = message.lower()
        return (
            "rate_limit" in m
            or "rate limit" in m
            or "too many requests" in m
            or "429" in m
        )

    @staticmethod
    def _friendly_daily(raw: str) -> str:
        hint = ""
        m2 = re.search(r"try again in\s+([^\n.]+)", raw, flags=re.I)
        if m2:
            hint = " Indicación de Groq: esperar ~" + m2.group(1).strip() + "."
        return (
            "Se alcanzó el límite diario de tokens (TPD) de Groq para este modelo." + hint
            + " Reduce uso (menos preguntas repetidas, menos historial en el chat), prueba otro modelo en GROQ_MODEL (.env), o revisa tu cuota en console.groq.com."
            + " [Detalle] " + raw
        )

    @staticmethod
    def _friendly_error(raw: str) -> str:
        if GroqClient._is_rate_limit(raw):
            wait = GroqClient._parse_retry_seconds(raw)
            if wait >= 1.0:
                mins = int(wait // 60)
                secs = int(wait % 60)
                wait_str = f"{mins}m {secs}s" if mins > 0 else f"{secs}s"
                return f"Límite de consultas Groq. Intentá de nuevo en {wait_str}."
            return "Límite de consultas Groq. Intentá de nuevo en unos minutos."
        return "Groq error: " + raw

    @staticmethod
    def _friendly_wait_long(wait_sec: float) -> str:
        mins = int(wait_sec // 60)
        secs = int(wait_sec % 60)
        wait_str = f"{mins}m {secs}s" if mins > 0 else f"{secs}s"
        return f"Límite de consultas Groq. Intentá de nuevo en {wait_str}."

    @classmethod
    def _get_wait_seconds(cls, exc: Exception, msg: str) -> float:
        try:
            ra = exc.response.headers.get("retry-after") or exc.response.headers.get("Retry-After")  # type: ignore[attr-defined]
            if ra:
                return min(300.0, max(1.0, float(ra)))
        except Exception:
            pass
        return cls._parse_retry_seconds(msg)

    @staticmethod
    def _parse_retry_seconds(message: str) -> float:
        m = re.search(r"try again in\s+(?:(\d+)h\s*)?(?:(\d+)m\s*)?([\d.]+)\s*s", message, flags=re.I)
        if m:
            h = float(m.group(1) or 0)
            mn = float(m.group(2) or 0)
            s = float(m.group(3) or 0)
            return min(300.0, max(0.5, h * 3600 + mn * 60 + s))
        return 3.0

    def _clamp_tool_json(self, json_str: str) -> str:
        if len(json_str) <= MAX_TOOL_JSON_BYTES:
            return json_str
        try:
            data = json.loads(json_str)
        except json.JSONDecodeError:
            return json_str
        if not isinstance(data, dict):
            return json_str
        for key in ("filas", "filas_ranking", "filas_pareto"):
            if key not in data or not isinstance(data[key], list):
                continue
            all_rows = data[key]
            for n in (12, 8, 5, 3):
                data[key] = all_rows[:n]
                data["_nota"] = f"Respuesta truncada: primeras {n} filas en {key} (límite API)."
                out = json.dumps(data, ensure_ascii=False, default=str)
                if len(out) <= MAX_TOOL_JSON_BYTES:
                    return out
            data[key] = all_rows
        return json.dumps(
            {
                "error": "respuesta_herramienta_muy_grande",
                "mensaje": "La consulta devolvió demasiados datos. Acote fechas o filtros.",
            },
            ensure_ascii=False,
        )
